### DISCUSSION TITLE
Using the COPC viewer to display data in public buckets?

### DISCUSSION THREAD

**Discussion by rsignell-usgs** [copcio/copcio.github.io#57] [Q&A]
We collected some point cloud data from the recent flooding in California and would like to allow stakeholders to visualize it using the COPC viewer.    We successfully converted our LAZ data to COPC using python-pdal and uploaded it to an Earth System Information Partners (ESIP) public bucket.   We were hoping this would then work:

https://viewer.copc.io/?copc=https://s3.amazonaws.com/esip-qhub-public/testing/20230105_MB_SC_Cap_UltraHi_NAD83_2011_UTM10N_ConfWater_Trimmed.copc.laz

But this just returns the error:  `Failed to fetch`. 

Is the viewer constrained to only load data from certain whitelisted buckets, to avoid abuse, perhaps?

This is our first time trying this also, so likely some user error, I guess.    Hopefully something simple!

**Comment by hobu** [comment]
https://s3.amazonaws.com/esip-qhub-public/testing/20230105_MB_SC_Cap_UltraHi_NAD83_2011_UTM10N_ConfWater_Trimmed.copc.laz is not publicly accessible.

**Comment by hobu** [comment]
> Is the viewer constrained to only load data from certain whitelisted buckets, to avoid abuse, perhaps?

You must make sure your bucket has viewer.copc.io listed in its CORS settings or it will not allow the fetch of data through it.

**Comment by rsignell-usgs** [comment]
@hobu, thanks, I thought you could slap that "s3.amazonaws.com" on the front of a standard s3 url:
`s3://esip-qhub-public/testing/20230105_MB_SC_Cap_UltraHi_NAD83_2011_UTM10N_ConfWater_Trimmed.copc.laz`  but I guess not.  

So the correct public https url to the COPC data should be:

https://esip-qhub-public.s3.amazonaws.com/testing/20230105_MB_SC_Cap_UltraHi_NAD83_2011_UTM10N_ConfWater_Trimmed.copc.laz

(not
https://s3.amazonaws.com/esip-qhub-public/testing/20230105_MB_SC_Cap_UltraHi_NAD83_2011_UTM10N_ConfWater_Trimmed.copc.laz)

I believe this viewer URL:

https://viewer.copc.io/?copc=https://esip-qhub-public.s3.amazonaws.com/testing/20230105_MB_SC_Cap_UltraHi_NAD83_2011_UTM10N_ConfWater_Trimmed.copc.laz

should work once CORS is configured correctly.  Does this look like the right CORS JSON policy for S3?
``` json
[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET",
            "HEAD"
        ],
        "AllowedOrigins": [
            "https://viewer.copc.io"
        ],
        "ExposeHeaders": []
    }
]
```
I tried the above URL for the viewer with this enabled policy and got the "failed to fetch" error.  But then I tried adding the COPC dataset in the GUI using the "+" button and now it works!
![2023-01-10_11-44-02](https://user-images.githubusercontent.com/1872600/211612331-8b9253db-4d8d-41a8-b5b9-f3236d0c9686.png)

**Comment by hobu** [comment]
Yep, looks good! https://viewer.copc.io?state=6158f9967c8bb8f21a80e8b872d00ab2a1437df913a46547f06867ab152b6551

**Comment by rsignell-usgs** [comment]
Thanks @hobu!   And just in case it's useful to folks who show up at this disussion, we converted our .laz to .copc.laz. using [this Jupyter Notebook](https://nbviewer.org/gist/7dcc5048acfaeef3cfc04ac1344d167d).

The conda environment we used to run the notebook was very simple:
``` yaml
name: pdal
channels:
- conda-forge
dependencies:
- python=3.10
- pdal
- python-pdal
- fsspec
- s3fs
- ipykernel
```

**Comment by hobu** [comment]
It is not required to reproject COPC data into EPSG:3857 to use the viewer. It should be able to consume most "boring" WKT data and do the right thing and reproject it as it loads it.

**Comment by rsignell-usgs** [comment]
@hobu, Okay!   Thanks for letting us know, since that was the most time consuming step in the workflow! I changed the notebook/gist to simplify it!
